worker/codalabworker/run_manager.py

```diff
@@ -0,0 +1,173 @@
+        """
+        raise NotImplementedError
+
+    @property
+    def cpus(self):
+        """
+        Total number of CPUs this RunManager has
+        """
+        raise NotImplementedError
+
+    @property
+    def gpus(self):
+        """
+        Total number of GPUs this RunManager has
+        """
+        raise NotImplementedError
+
+    @property
+    def memory_bytes(self):
+        """
+        Total installed memory of this RunManager
+        """
+        raise NotImplementedError
+
+class Reader(object):
+    def __init__(self):
+        pass
+
+    def read(self, run_state, path, dep_paths, read_args, reply):
+        bundle_uuid = run_state.bundle['uuid']
+        dep_paths = set([dep['child_path'] for dep in run_state.bundle['dependencies']])
+        read_type = read_args['type']
+        if read_type == 'get_target_info':
+            self.get_target_info(run_state, path, dep_paths, read_args, reply)
+        elif read_type == 'stream_directory':
+            self.stream_directory(run_state, path, dep_paths, read_args, reply)
+        elif read_type == 'stream_file':
+            self.stream_directory(run_state, path, dep_paths, read_args, reply)
+        elif read_type == 'read_file_section':
+            self.read_file_section(run_state, path, dep_paths, read_args, reply)
+        elif read_type == 'summarize_file':
+            self.summarize_file(run_state, path, dep_paths, read_args, reply)
+        else:
+            err = (httplib.BAD_REQUEST, "Unsupported read_type for read: %s" % read_type)
+            reply(err)
+
+    def get_target_info(self, run_state, path, dep_paths, args, reply_fn):
+        """
+        Calls reply_fn(err, msg, data) with the target_info of the path
+        in the msg field, or with err if there is an error
+        """
+        raise NotImplementedError
+
+    def stream_directory(self, run_state, path, dep_paths, args, reply_fn):
+        """
+        Calls reply_fn(err, msg, data) with the dir contents of the path
+        in the data field, or with err if there is an error
+        """
+        raise NotImplementedError
+
+    def stream_file(self, run_state, path, dep_paths, args, reply_fn):
+        """
+        Calls reply_fn(err, msg, data) with the file contents of the path
+        in the data field, or with err if there is an error
+        """
+        raise NotImplementedError
+
+    def read_file_section(self, run_state, path, dep_paths, args, reply_fn):
+        """
+        Calls reply_fn(err, msg, data) with the file section contents of the path
+        in the data field, or with err if there is an error
+        """
+        raise NotImplementedError
+
+    def summarize_file(self, run_state, path, dep_paths, args, reply_fn):
+        """
+        Calls reply_fn(err, msg, data) with the file summary of the path
+        in the data field, or with err if there is an error
+        """
+        raise NotImplementedError
+
+class RunState(object):
```

Comment:
does this get serialized to json properly?

Reply:
Good point. The reason I did this was thinking that different architectures might want to store different things in RunStates, but we might want all RunStates to have a base minimum set of fields, I was hoping the Class would make it more extensible. I have to look into serialization. Maybe having it inherit from Dict instead of object could help.